core/arch/arm/kernel/user_ta.c

```diff
@@ -621,21 +504,410 @@ TEE_Result tee_ta_register_ta_store(struct user_ta_store_ops *ops)
+	size_t i;
+
+	if (strlen(p) != 36)
+		return TEE_ERROR_BAD_FORMAT;
+	if (p[8] != '-' || p[13] != '-' || p[18] != '-' || p[23] != '-')
+		return TEE_ERROR_BAD_FORMAT;
+
+	u.timeLow = parse_hex(p, 8, &res);
+	if (res)
+		goto out;
+	p += 9;
+	u.timeMid = parse_hex(p, 4, &res);
+	if (res)
+		goto out;
+	p += 5;
+	u.timeHiAndVersion = parse_hex(p, 4, &res);
+	if (res)
+		goto out;
+	p += 5;
+	for (i = 0; i < 8; i++) {
+		u.clockSeqAndNode[i] = parse_hex(p, 2, &res);
+		if (res)
+			goto out;
+		if (i == 1)
+			p += 3;
+		else
+			p += 2;
+	}
+	*uuid = u;
+out:
+	return res;
+}
+
+static TEE_Result add_elf_deps(struct user_ta_ctx *utc, char **deps,
+			       size_t num_deps)
+{
+	struct user_ta_elf *libelf;
+	TEE_Result res = TEE_SUCCESS;
+	TEE_UUID u;
+	size_t n;
+
+	for (n = 0; n < num_deps; n++) {
+		res = parse_uuid(deps[n], &u);
+		if (res) {
+			EMSG("Invalid dependency (not a UUID): %s", deps[n]);
+			goto out;
+		}
+		DMSG("Library needed: %pUl", (void *)&u);
+		libelf = ta_elf(&u, utc);
+		if (!libelf) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			goto out;
+		}
+	}
+out:
+	return res;
+}
+
+static TEE_Result resolve_symbol(struct user_ta_elf_head *elfs,
+				 const char *name, uintptr_t *val)
+{
+	struct user_ta_elf *elf;
+	TEE_Result res;
+
+	/*
+	 * The loop naturally implements a breadth first search due to the
+	 * order in which the libraries were added.
+	 */
+	TAILQ_FOREACH(elf, elfs, link) {
+		res = elf_resolve_symbol(elf->elf_state, name, val);
+		if (res == TEE_ERROR_ITEM_NOT_FOUND)
+			continue;
+		if (res)
+			return res;
+		*val += elf->load_addr;
+		FMSG("%pUl/0x%" PRIxPTR " %s", (void *)&elf->uuid, *val, name);
+		return TEE_SUCCESS;
+	}
+
+	return TEE_ERROR_ITEM_NOT_FOUND;
+}
+
+static TEE_Result load_elf_from_store(const TEE_UUID *uuid,
+				      const struct user_ta_store_ops *ta_store,
+				      struct user_ta_ctx *utc)
+{
+	struct user_ta_store_handle *handle = NULL;
+	struct elf_load_state *elf_state = NULL;
+	struct ta_head *ta_head;
+	struct user_ta_elf *elf;
+	char **deps = NULL;
+	size_t num_deps = 0;
+	TEE_Result res;
+	size_t vasize;
+	void *p;
+	size_t n;
+	size_t num_segs = 0;
+	struct load_seg *segs = NULL;
+
+	res = ta_store->open(uuid, &handle);
+	if (res)
+		return res;
+
+	elf = ta_elf(uuid, utc);
+	if (!elf) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	res = elf_load_init(ta_store, handle, elf->is_main, &utc->elfs,
+			    resolve_symbol, &elf_state);
+	if (res)
+		goto out;
+	elf->elf_state = elf_state;
+
+	res = elf_load_head(elf_state,
+			    elf->is_main ? sizeof(struct ta_head) : 0,
+			    &p, &vasize, &utc->is_32bit);
+	if (res)
+		goto out;
+	ta_head = p;
+
+
+	elf->mobj_code = alloc_ta_mem(vasize);
+	if (!elf->mobj_code) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+	/*
+	 * Zero initialize everything to make sure that all memory not
+	 * updated when loading the ELF is zero (covering .bss and eventual
+	 * gaps)
+	 */
+	memset(mobj_get_va(elf->mobj_code, 0), 0, vasize);
```

Comment:
This doesn't work with pager. On the other hand this is not needed with paged memory, perhaps this should be done conditionally in `alloc_ta_mem()` instead.